Added diagnostic check for watertight nodesets

DanielYankura opened the pull request:
-Loops through all elements and sides
-If side is external it checks if all its nodes are assigned to a nodeset

closes #28822

<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->

## Reason
<!--Why do you need this feature or what is the enhancement?-->
Adds new diagnostic tool for users

## Design
<!--A concise description (design) of the enhancement.-->
If feature it turned on, it loops through each element, then the element's sides. If the sides are external it checks each node to ensure it's been assigned to a nodeset. If not, the node id is printed for reference.

## Impact
<!--Will the enhancement change existing APIs or add something new?-->
Adds watertight nodeset check to Mesh Diagnostics 

<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->

DanielYankura committed:
Added diagnostic check for watertight nodesets

-Loops through all elements and sides
-If side is external it checks if all its nodes are assigned to a nodeset

closes #28822

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -45,6 +45,10 @@ MeshDiagnosticsGenerator::validParams()
+      "check_for_watertight_nodesets",
+      chk_option,
+      "whether to check for external nodes that are not assigned to any nodeset");
   params.addParam<MooseEnum>(
```
let s add a parameter for the nodesets that are supposed to englobe the entire mesh. when this parameter is in use, the nodes are only checked against those "blessed" nodesets

same for sidsets
actually now that I see that this is not an option for sidesets, let s keep it for future work

DanielYankura replied:
Sounds good. I'll open a new issue for adding that parameter to the sideset and nodeset checks.

DanielYankura committed:
Made formatting changes for Civet

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -335,6 +343,65 @@ MeshDiagnosticsGenerator::checkWaterTightSidesets(const std::unique_ptr<MeshBase
+    mooseError("The nodeset check only works for 2D and 3D meshes");
+  auto & boundary_info = mesh->get_boundary_info();
+  boundary_info.build_side_list();
+  const auto nodeset_map = boundary_info.get_nodeset_map();
```
where is the _boundary_node_id being initialized in libmesh? is the previous call doing it?

DanielYankura replied:
So I looked into it. The node ids get added to _boundary_node_id after execute_mesh_generators is called in ActionWarehouse.C. From what I can tell the mesh is always initialized before this test could ever be called.

GiudGiud replied:
>  after execute_mesh_generators

wouldnt that be too late then?

DanielYankura replied:
Sorry, it's the execute_mesh_generators task that initializes _boundary_node_id. After initialization mesh diagnostics runs. So that task runs both I believe, but it should initialize first.

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -335,6 +343,65 @@ MeshDiagnosticsGenerator::checkWaterTightSidesets(const std::unique_ptr<MeshBase
+
+  for (const auto elem : mesh->active_element_ptr_range())
+  {
+    for (auto i : elem->side_index_range())
```
```suggestion
    for (const auto i : elem->side_index_range())
```

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -335,6 +343,65 @@ MeshDiagnosticsGenerator::checkWaterTightSidesets(const std::unique_ptr<MeshBase
+      // Check if side is external
+      if (elem->neighbor_ptr(i) == nullptr)
+      {
+        // Side is externa, now check nodes
```
```suggestion
        // Side is external, now check nodes
```

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -335,6 +343,65 @@ MeshDiagnosticsGenerator::checkWaterTightSidesets(const std::unique_ptr<MeshBase
+            if (num_nodes_without_nodeset < _num_outputs)
+            {
+              message = "Node " + std::to_string(node->id()) +
+                        " is external, but has not been assigned to a nodeset";
```
```suggestion
                        " is on an external boundary of the mesh, but has not been assigned to a nodeset";
```

GiudGiud commented on test/tests/meshgenerators/mesh_diagnostics_generator/tests:
```diff
@@ -45,6 +45,22 @@
+      cli_args = '--mesh-only'
+      mesh_mode = 'replicated'
+      expect_out = 'Number of external nodes that have not been assigned to a nodeset: 1'
+      detail = '2D meshed square missing a nodeset on one side'
```
```suggestion
      detail = '2D meshed square with one external node that does not belong to a nodeset'
```

GiudGiud commented on test/tests/meshgenerators/mesh_diagnostics_generator/tests:
```diff
@@ -45,6 +45,22 @@
+      cli_args = '--mesh-only'
+      mesh_mode = 'replicated'
+      expect_out = 'Number of external nodes that have not been assigned to a nodeset: 1'
+      detail = '3D meshed cube missing a nodeset on one side'
```
same

GiudGiud reviewed (commented):
looks good

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2546963/), step Docs: sync website on 0a1493f wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/29068/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2546978/), step Generate coverage on 0a1493f wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>737741</th>
      <th colspan=4>#29068 0a1493</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.11%</td>
      <td>85.11%</td>
      <td>+0.00%</td>
      <td>97.56%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>106985</td>
      <td>107021</td>
      <td>+36</td>
      <td>40</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18718</td>
      <td>18719</td>
      <td>+1</td>
      <td>1</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/29068/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/29068/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/29068/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/29068/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/29068/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/29068/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/29068/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/29068/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/29068/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/29068/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/29068/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/29068/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/29068/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/29068/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/29068/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/29068/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/29068/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/29068/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/29068/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/29068/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/29068/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/29068/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/29068/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/29068/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/29068/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/29068/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/29068/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/29068/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/29068/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/29068/coverage/xfem)

</details>



This comment will be updated on new commits.

DanielYankura committed:
Made changes for PR

-Fixed spelling mistake
-made iterator const
-improved error messages

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2548505/), step Clang format on 0c3d1cd wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/29068/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/29068/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7377410027ba9dbf48902235a0437fdf96485e29`

DanielYankura committed:
Style changes for Civte again

DanielYankura commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -335,6 +343,66 @@ MeshDiagnosticsGenerator::checkWaterTightSidesets(const std::unique_ptr<MeshBase
+          {
+            // This node does not have a nodeset!!!
+            num_nodes_without_nodeset++;
+            checked_nodes.insert(*node);
```
Thanks! I was trying to figure out why it was failing that test. I think I have the same issue in #28701 so I'll fix it in that one as well.

GiudGiud replied:
I didnt remember that one was still open. Do you need anything else from me for that? Are we waiting on Roy's review?

DanielYankura replied:
I think I'm just waiting on Roy's review. Though if this one get's merged first I might have to make some changes to the other before it can be merged.

DanielYankura committed:
checked_nodes array now takes a node id instead of a node to avoid libmesh deprecation error

roystgnr commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -335,6 +343,66 @@ MeshDiagnosticsGenerator::checkWaterTightSidesets(const std::unique_ptr<MeshBase
+  if (mesh->mesh_dimension() < 2)
+    mooseError("The nodeset check only works for 2D and 3D meshes");
+  auto & boundary_info = mesh->get_boundary_info();
+  boundary_info.build_side_list();
```
What's going on here?  This is the `build_side_list()` variant that returns a vector of tuples, but if nothing uses the return value then it's just an expensive no-op.

DanielYankura replied:
I think I copied that over from a similar function I made previously and then forgot to remove it. I'l fix that real quick.

roystgnr commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -335,6 +343,66 @@ MeshDiagnosticsGenerator::checkWaterTightSidesets(const std::unique_ptr<MeshBase
+          if (checked_nodes_id.count(node->id()))
+            continue;
+          // if node is not part of nodeset map add it to list of bad nodes
+          if (nodeset_map.count(node) == 0)
```
Could we just test `boundary_info.n_boundary_ids(node)` here?  The raw `get_nodeset_map()` accessor was an encapsulation mistake I should never have approved, and I still vaguely dream of deprecating it someday.

DanielYankura replied:
Just want to check, would that return true if the node is assigned to a sideset but not a nodeset?

roystgnr reviewed (changes requested):
In principle, I oppose the very concept of checking for watertight nodesets, and would be happy to rant at length about why.  (nobody else would be happy, so I won't)

In practice, I recognize that some MOOSE workflows will be using nodesets as pseudo-sidesets until the end of time; this is a great way for users of those workflows to check for meshing problems.  I just have one minor problem and one minor request here.

DanielYankura committed:
made changes for PR

-removed unnecessary initialization of side_list
-removed call to nodeset_map and instead used boundary_info.n_boundary_ids(node)
